sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java

```diff
@@ -431,7 +431,20 @@ public static GenericRecord toGenericRecord(
    */
   public static SerializableFunction<Row, GenericRecord> getRowToGenericRecordFunction(
       @Nullable org.apache.avro.Schema avroSchema) {
-    return g -> toGenericRecord(g, avroSchema);
+    return new RowToGenericRecordFn(avroSchema);
+  }
+
+  private static class RowToGenericRecordFn implements SerializableFunction<Row, GenericRecord> {
+    private final transient org.apache.avro.Schema avroSchema;
+
+    RowToGenericRecordFn(@Nullable org.apache.avro.Schema avroSchema) {
+      this.avroSchema = avroSchema;
+    }
+
+    @Override
+    public GenericRecord apply(Row input) {
+      return toGenericRecord(input, avroSchema);
+    }
```

Comment:
Couldn't you just implement `equals` here instead of changing to comparing byte equality of the serialized function in SchemaCoder?

Reply:
This is a second different issue about capture of Avro schema on serialization so not really related to equals. As explained above I put both together because I use equality to validate the roundtrip of serialization/deserialization.